### PR TITLE
expr: remove matching on Datum

### PR DESCRIPTION
**Pull request by benesch**
Explicitly matching on the type of Datum inside of the expr crate
indicates that the SQL type system has failed. (All type matching should
happen implicitly through function selection.) Refactor away a few cases
that had slipped through in table functions.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3697)

### CODE REVIEW

**Comment by frankmcsherry** [comment]
I feel like this is a direction we should talk about at some point. It seems to be moving away from "errors we can surface" and towards forcing panics on the system. If that is the way we want to build things then fair enough, but I don't think we concluded that (in a discussion with @awang on stability). If we did, my apologies!

**Comment by frankmcsherry** [comment]
We might want to close down #126 and #1177, for example.

**Comment by benesch** [comment]
> I feel like this is a direction we should talk about at some point. It seems to be moving away from "errors we can surface" and towards forcing panics on the system. 

This PR is just enforcing an invariant that has existed since the very beginning of the `expr` crate: https://github.com/MaterializeInc/materialize/commit/efb2bbb1763a087648b9d4d8cb9dc36f79df177d

I'm definitely down to talk about this philosophy—in fact I wrote up a very long comment with my thoughts in #3624. But no matter what I would very much like to remove these instances of "function silently returns empty/null when type is wrong", because when that goes wrong it is way, way harder to track down than an error or a panic.

For example one outcome of #3624 might be that we should replace

```
    pub fn unwrap_int64(&self) -> i64 {
        match self {
            Datum::Int64(i) => *i,
            _ => panic!("Datum::unwrap_int64 called on {:?}", self),
        }
    }
```

with:

```
    pub fn try_get_int64(&self) -> Result<i64, EvalError> {
        match self {
            Datum::Int64(i) => *i,
            _ => EvalError::ProgrammingError(format!("type mismatch: expected i64 but got {:?}", self)),
        }
    }
```

Much easier to apply that search/replace if the `expr` crate is not matching on datums directly.

All this to say: I'd appreciate your thoughts on #3624!

File: src/expr/src/relation/func.rs [hunk 1]
```diff
@@ -543,33 +543,32 @@ fn jsonb_array_elements<'a>(a: Datum<'a>, temp_storage: &'a RowArena, stringify:
...
-            let datums_iter = captures
-                .iter()
-                .skip(1)
-                .map(|m| Datum::from(m.map(|m| m.as_str())));
-            Some(row_packer.pack(datums_iter))
-        }
-        _ => None,
-    }
+    let r = r.inner();
+    let captures = r.captures(a.unwrap_str())?;
```
Comment (umanwizard) [inline src/expr/src/relation/func.rs:547]: Stylistically I’d slightly prefer the unwrap to be on its own line (`let s = a.unwrap_str()`), on the principle that “panics should jump out at you while skimming”

File: src/expr/src/relation/func.rs [hunk 2]
```diff
@@ -651,28 +650,19 @@ impl AnalyzedRegex {
...
-    let bytes = match a {
-        Datum::Bytes(bytes) => bytes,
-        Datum::String(str) => str.as_bytes(),
-        _ => return vec![],
-    };
+    let mut row_packer = RowPacker::new();
     let mut csv_reader = csv::ReaderBuilder::new()
         .has_headers(false)
-        .from_reader(bytes);
+        .from_reader(a.unwrap_str().as_bytes());
```
Comment (umanwizard) [inline src/expr/src/relation/func.rs:658]: Idem

File: src/expr/src/relation/func.rs (hunk 1 above)
Comment (benesch) [reply src/expr/src/relation/func.rs]: 👍 done

File: src/expr/src/relation/func.rs (hunk 2 above)
Comment (benesch) [reply src/expr/src/relation/func.rs]: 👍 also done

**Pull request merged**